### ISSUE TITLE
Producer/Consumer pattern for furniture

### ISSUE THREAD

**Issue by Grenkin1988** [TeamPorcupine/ProjectPorcupine#604]
Recently I was think of some generic pattern for our furniture so it could handle complicated stuff:
-   Some furniture could produce multiple resources
-   Some furniture could consume multiple resources
-   Some furniture could produce multiple things while consuming another

My thought was to have something like:

```
<Consume>
    <Item name=”” amount=””/>
</Consume>
<Produce>
    <Item name=”” amount=””/>
</ Produce >
```

Some thoughts were here #602
What do you think about this?

**Comment by Xeni91** [comment]
I think it is good but we need to implement a time aspekt in this. Consumption and Production can be after every operation, but when some furniture needs a constant supply of power/gas/oxygen we should be able to set the intervall.

**Comment by Grenkin1988** [comment]
Right now we need to gather all the requirements for this, and decide whether this is good idea at all

**Comment by mikejbrown** [comment]
So `<Item name=”” amount=”” rate=""/>` instead?

How about this calling arrangement:

- Producer OnUpdate makes X units of resource Y & calls a callback to a consumer registered for listening to that resource (scheduling could be an issue if there are >1 consumers?), essentially giving an _offer_ to the consumer...
- The Consumer, in its callback, decides whether or not it wants to use that resource & how much. Then it calls a function on the producer saying "I'm taking Z much of Y now" so...
- The Producer can now update its state, offer any remaining resource to the next consumer etc...

Just quickly sketching ideas here. Thoughts?

**Comment by dusho** [comment]
It definitely makes sense to make some generic factory. From the looks at Metal Smelter it should cover most of LUA code ( together with that stacking code - it shouldn't be in LUA at all)
Would this cover all that is needed?:
```
<Factory processingSpeed="">
	<Input>
		<FactoryItem name="" amount="" position=""></FactoryItem>
		<FactoryItem name="" amount="" position=""></FactoryItem>
	</Input>
	<Output>
		<FactoryItem name="" amount="" position=""></FactoryItem>
		<FactoryItem name="" amount="" position=""></FactoryItem>
	</Output>
</Factory>
```
where position would be spawn points (required position for input, spawn position for output), and processingSpeed would be rate of taking all items from Input and producing Output. If Output is full (can't stack anymore) or Input is not fulfilled, factory will not produce anything.

**Comment by Grenkin1988** [comment]
I would definitely go for making derived class for `Factory`

**Comment by RodRitter** [comment]
I think we could change FactoryItem to InventoryItem to fit #597

**Comment by dusho** [comment]
I would still create FactoryItem, it should inherit from InventoryItem then and add new member ÍntVector2d relSpawnPosition; // relative placement to spawn position

**Comment by bjubes** [comment]
keep in mind two types of production. a smelter produces differenly (5 iron -> five plates, every two seconds) than a o2 generator(power -> o2 at a given conversion rate and production time)

**Comment by ajahraus** [comment]
Here's a weird question, are characters producers/consumers? Like, if a character is working at a station, does that work count in what's consumed during the job? Or is 'Work' handled in a different system, and if so, should it be unified with is system? 

I'm not casting a vote, I'm just wondering if this is worth considering.

**Comment by mikejbrown** [comment]
@ajahraus Am I understanding you right that you're suggesting a character produces a "resource" called "work" (or whatever) which is consumed by stations? I think that's an interesting idea. It gives an easy way to build in character leveling: they just produce work units more quickly. Then there is no need to tweak station stats or job stats or anything like that.

**Comment by ajahraus** [comment]
@mikejbrown Yeah, that's a good way to think of it. Labour as a resource is also a good way to think about personal time management IRL, so that might also play well into the GOAP being used in character AI, #86. Since each character has to decide what to do with their own time, deciding on a task may just be as simple as maximizing the 'value' for the resource 'Work.' 

That also has some implications for personality traits. Like, a 'Hard Worker' over-values their 'Work,' leaving them productive but in a poor state of physical well-being. Or a 'Slob' under values cleaning. But perhaps this line of thinking belongs in a different thread.

**Comment by mikejbrown** [comment]
@ajahraus I really like that idea, but it possibly should wait for a general GOAP implementation (or be done as a part of). It might be too much for this particular issue.

**Comment by kd7uiy** [comment]
Let's say that we have a factory. That factory can produce Steel from Raw Iron, and Copper from Raw Copper. We should set up the system so that can be done. It's more like:

```
<Item>
    <Produce item="" amount="">
    <Consume item="" amount=""
</Item>
<Item>
    <Produce item="" amount="">
    <Consume item="" amount=""
</Item>
```

**Comment by Grenkin1988** [comment]
This could be simpler
'''
<Produce item="" amount="">
       <Consume item="" amount=""/>
       <Consume ..... />
</Produce>
'''

**Comment by dusho** [comment]
@Grenkin1988 can't see what you put there.. just blank
will try to implement this (with metal smelter and power cell press) as WIP and see how it goes

**Comment by dusho** [comment]
So there is WIP now ( #1084 ).
Things l like to know if it's cool:
- is it cool to use partial classes that will contain specific implementation ? e.g. I introduced Furniture.Factory.cs that holds stuff related to furniture consuming/producing stuff
- my WIP should allow to have multiple inputs (even multiple outputs), so to present it, I would like to recreate Power cell press furniture to have 2 inputs (Steel plates 2x, uranium 1x) and produce power cell 1x - then power cell can be used to power machines? instead of raw uranium ? is that cool?
- previous system was 'waiting' for some time and then at same time as output popped out, input was decreased, I changed it so that it works in phases -> input is taken, processing started (param flag is set), processing time passes, output is spit out (if there is space) -> something like Factorio uses
